### ISSUE TITLE
Frontend bug: profit disappears after selling all assets

### ISSUE THREAD

**Issue by sagata-1** [sagata-1/marketsdojo_backend#12]
There is an inconsistency on the reported profit when one performs a sequence of buys and sells on an asset. 

Some relevant chat messages reproduced below for reference:

2/16 Pranav:
> Marketsdojo.com live code is buggy I think problem on frontend side …you can see for user fridayevening@friday.com, password: test….I have sold all assets in his portfolio and made a profit of $51.79 but the profit/loss column shows $0. You can replicate the situation I’m talking about by buying some assets then selling them and seeing what happens.
> 
> Then if I buy—>sell—>buy in the invested column it doesn’t show the latest invested amount it is showing the amount invested on the first buy.

2/17 Pranav:
> @⁨Arman Jasuja⁩ hi arman. I believe this is related to buy/sell logic do you want to take a look? If I buy a stock for $100, then my invested amount is $100…if I sell it then my invested amount is $0 cause I’m not invested in it…so then if I buy it again and the price is $200 at that time my invested amount should show $200. Currently it’s showing $100

2/17 Arman:
> Yes… that’s that big complex pay logic we discussed, and you said we can stick to the basic version for now. Because otherwise we’d have to keep track of price at first buy, second buy, third buy, etc…. currently, when selling invested amount, it sells of all the amount at market price, but for the actual correct amount, it would need to sell off at all the bought prices, by precisely that amount
>
> And for multiple buy’s would have to keep track of each buy, and amount bought at each price, so can sell of exactly that amount

**Comment by tedz2usa** [comment]
Right now, the backend restricts buy/sell activity to when the markets are open. I propose we introduce some developer diagnostics mode that allows us to test buy/sell activity no matter whether the markets are open or not. 

I also propose, for reproducibility, in the diagnostics mode, we use a mock price tick history sequence that is always the same. That way when we try the same sequence of buy/sell actions, the result is the same. This will allow us to positively reference known bugs, and positively confirm when the bug has been fixed.

**Comment by sagata-1** [comment]
@tedz2usa it was determined by @armurox and I that the backend is not the source of the problem...especially since the profit/loss is not even calculated on the backend at the moment.

**Comment by armurox** [comment]
> Right now, the backend restricts buy/sell activity to when the markets are open. I propose we introduce some developer diagnostics mode that allows us to test buy/sell activity no matter whether the markets are open or not.
> 
> I also propose, for reproducibility, in the diagnostics mode, we use a mock price tick history sequence that is always the same. That way when we try the same sequence of buy/sell actions, the result is the same. This will allow us to positively reference known bugs, and positively confirm when the bug has been fixed.

@tedz2usa This makes sense. I suggest that we create a specific user that has this ability, and allow him to execute buy / sell to occur without time restriction with mock prices when he is logged in.

@sagata-1 I think the problem is, since he can’t execute any buy / sells from the front end, he can’t really test the issue at all anyway (additionally, more generally, we can’t ever claim that the backend is free of bugs, just that the specific bug referenced in the chats has been checked on a pure backend basis and confirmed to not to be the issue). It could be the case that what the front-end is using to compute profit and loss is incorrect on the backend. I am quite busy this week, but can look at doing this backend conversion maybe on the weekend if this is ok.

**Comment by sagata-1** [comment]
@armurox I know what the problem is, he and I had a meeting before your comment above, no clarification needed here thanks. " just that the specific bug referenced in the chats has been checked on a pure backend basis and confirmed to not to be the issue" - is exactly what I said. There's a specific issue that really seems like a front-end issue if the price of GooG was never $136 on the day the user bought it but still it shows to be $136. I believe it is possible to make the claim that the backend is free of bugs. If the backend had only 1 function and if p-->q evaluates to true for that function, then you can make the claim that the function is correct i.e. free of bugs..so on so forth for every function in the backend will lead to the claim that the backend is free of bugs. For all strings where |s|>0 remove the last letter...if you write the precondition and postcondition for the function before implementing then implement it to ensure the post-condition is true whenever the precondition is true, the claim that the function is correct and free of bugs would be a true statement.

**Comment by tedz2usa** [comment]
It looks like we are talking about two different issues:
 
A. The Profit/Loss value is incorrect after selling all assets of a particular symbol.
B. In the portfolio listing, the Invested column value is incorrect.

### Regarding A: The Profit/Loss value is incorrect after selling all assets of a particular symbol
The reason for this is currently, the Profit/Loss value that is reported is only calculated on currently owned assets. Therefore, in the current state of the application when you sell all AAPL shares, even if you made a profit, that net profit no longer shows up in the Profit/Loss field because the Profit/Loss field only shows the net for currently owned assets.

The Profit/Loss field is calculated on the frontend. We would want to confirm the exact specification for the formula for this value, but essentially, we may want to change it to: 

```
Profit_Loss = (Available_Balance + Current Value) - Opening_Balance
```

This would encompass profit made from assets that were fully sold off.

### Regarding B: In the portfolio listing, the Invested column value is incorrect
This number is reported by the backend API `/v1/users/portfolio`, where the `invested_amount` value is calculated by doing a summation of the `invested_amount_per_transaction` column in the history table for transactions with the given symbol for the given user (implemented by function `calculate_portfolio_invested_amount`). This appears to be the correct calculation. 

I manually checked the historical transactions of the `fridayevening@friday.com` user, and it appears to be correctly adding up. I don't see indication of behavior where, say, a purchase of $100 of AAPL, followed by a sell of $100 AAPL,  followed by a buy of $200 AAPL shows $100. The summation should show the invested amount correctly at $200. 
It would be great if we can isolate a reproducible case of this bad behavior from a sequence of specific transactions (the diagnostics mode may help with this type of debugging).

**Comment by armurox** [comment]
> It looks like we are talking about two different issues:
> 
>  
> 
> A. The Profit/Loss value is incorrect after selling all assets of a particular symbol.
> 
> B. In the portfolio listing, the Invested column value is incorrect.
> 
> 
> 
> ### Regarding A: The Profit/Loss value is incorrect after selling all assets of a particular symbol
> 
> The reason for this is currently, the Profit/Loss value that is reported is only calculated on currently owned assets. Therefore, in the current state of the application when you sell all AAPL shares, even if you made a profit, that net profit no longer shows up in the Profit/Loss field because the Profit/Loss field only shows the net for currently owned assets.
> 
> 
> 
> The Profit/Loss field is calculated on the frontend. We would want to confirm the exact specification for the formula for this value, but essentially, we may want to change it to: 
> 
> 
> 
> ```
> 
> Profit_Loss = (Available_Balance + Current Value) - Opening_Balance
> 
> ```
> 
> 
> 
> This would encompass profit made from assets that were fully sold off.
> 
> 
> 
> ### Regarding B: In the portfolio listing, the Invested column value is incorrect
> 
> This number is reported by the backend API `/v1/users/portfolio`, where the `invested_amount` value is calculated by doing a summation of the `invested_amount_per_transaction` column in the history table for transactions with the given symbol for the given user (implemented by function `calculate_portfolio_invested_amount`). This appears to be the correct calculation. 
> 
> 
> 
> I manually checked the historical transactions of the `fridayevening@friday.com` user, and it appears to be correctly adding up. I don't see indication of behavior where, say, a purchase of $100 of AAPL, followed by a sell of $100 AAPL,  followed by a buy of $200 AAPL shows $100. The summation should show the invested amount correctly at $200. 
> 
> It would be great if we can isolate a reproducible case of this bad behavior from a sequence of specific transactions (the diagnostics mode may help with this type of debugging).
> 
> 

I agree, thank you for clarifying. In terms of  the backend check I did, it was indeed with respect to B, and that was also what I was referring to when agreeing with the admin user set-up. 

### Regarding A:

The claim I was making was that it was possible that available balance was being incorrectly computed on the backend or some other value which the front end was dependant on for its P&L calculation, which has not been checked specifically. (I.e. non-obvious backend bugs which may be affecting the front-end)

**Comment by tedz2usa** [comment]
@armurox Absolutely, I saw you guys were mainly talking about Issue B, I understood your backend check was for that issue.
